fix: Solves Issues with Set Function in Component Class and Enhance Input Handling

edwinjosechittilappilly opened the pull request:
This pull request addresses improvements in the `Component` class's input handling and introduces new unit tests to ensure robust functionality. The following changes have been made:

1. **Improved Input Processing**:
   - Updated the `_process_connection_or_parameters` method in `component.py` to ensure that when a value is a list, it correctly processes only lists of `Component` instances. This enhancement prevents potential errors when other data types are passed.

2. **New Unit Tests**:
   - Added test cases to verify the behavior of the `set` method in the `Component` class:
     - **Mixed List Input Test**: Checks that the `set` method can handle a list containing both a string and a component.
     - **MessageTextInput List Test**: Ensures that the `set` method can accept a list of `MessageTextInput` instances and that their values are set correctly.
     - **Single Component Input Test**: Validates that a single component can be set correctly.

### Code Changes:
- **Component Logic**:
  ```python
  def _process_connection_or_parameters(self, key, value) -> None:
      # if value is a list of components, we need to process each component
      # Note this update makes sure it is a list of components and not of any other values
      if isinstance(value, list) and all(isinstance(item, Component) for item in value):
          for val in value:
              self._process_connection_or_parameter(key, val)
      else:
          self._process_connection_or_parameter(key, value)
  ```

- **New Tests**:
  - Tests added in `test_component_set_functionality.py` to cover various input scenarios.

### Notes:
- This PR solves the inconsistency in input types when the component was run as a tool, enhancing the reliability and flexibility of the `Component` class by ensuring it can handle diverse input types effectively.

ogabrielluiz reviewed (approved):
LGTM

jordanrfrazier commented on src/backend/base/langflow/custom/custom_component/component.py:
```diff
@@ -442,7 +442,8 @@ def _process_connection_or_parameter(self, key, value) -> None:
         # if value is a list of components, we need to process each component
-        if isinstance(value, list):
+        # Note this update make sure it is a list of components and not of any other values
+        if isinstance(value, list) and all(isinstance(item, Component) for item in value):
```
This feels like it's missing a case. We have:
a) IF `value` is a list AND they're all `Components`, process each item in the list separately. 
b) ELSE, process the input as a single value. 

What happens if it's a list, but they're not all `Components`?

edwinjosechittilappilly replied:
it executes else and assigns the values as per key value pair

edwinjosechittilappilly replied:
I am not sure if we will face that scenario if so we could add a case for it change the case.

edwinjosechittilappilly replied:
@jordanrfrazier  you were right in a way. I guess we should add multiple cases. Added a case just now to solve the pytest failure error.

edwinjosechittilappilly commented:
In Pytest: test_starter_projects that tests 
api/v1/starter-projects/
Am getting an Error:  {"message":"[\"Something went wrong while serializing the response. Please share this error on our GitHub repository.\", \"Unable to serialize unknown type: <class 'method'>\"]"}
Any suggestions: @ogabrielluiz @jordanrfrazier @italojohnny ?

edwinjosechittilappilly commented:
@jordanrfrazier  I have added a case where all the inputs are not components and updated the function accordingly.

Even though PR#4269 solves serialization error.
The function _ process_connection_or_parameters also caused some of those errors; hence, it was updated to solve the serialisation of values in _process_connection_or_parameters.

Tested with pytest and all cases passes.
